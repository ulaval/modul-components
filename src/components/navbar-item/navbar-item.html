<li class="m-navbar-item"
    :class="{'m--is-selected': isSelected,
             'm--is-disabled': isDisabled}"
    role="menuitem"
    :aria-selected="isSelected"
    @keyup.enter="onClick"
    @click="onClick"
    @mouseover="onMouseover"
    @mouseleave="onMouseleave">
    <slot></slot>
</li>
