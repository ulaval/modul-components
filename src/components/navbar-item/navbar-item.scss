@import 'abstracts/variables';

a {
    text-decoration: none;
}

.m--is-skin-regular {
    .m-navbar-item {
        + .m-navbar-item {
            margin-left: 1px;
        }

        &:not(.m--is-disabled) {
            &[tabindex],
            a,
            button {
                padding: $m-padding;
                background: $m-color--grey-light;
                color: $m-color--text;
            }
        }

        &.m--is-selected {
            position: relative;

            &[tabindex],
            a,
            button {
                background: $m-color--white;

                &::before {
                    position: absolute;
                    top: 0;
                    right: 0;
                    left: 0;
                    content: '';
                    height: $m-border-width--l;
                    background: $m-color--active;
                }
            }

            &.m--is-first {
                &::after {
                    position: absolute;
                    top: $m-border-width--l;
                    bottom: 0;
                    left: 0;
                    content: '';
                    width: $m-border-width--s;
                    background: $m-color--grey;
                }
            }

            &.m--is-last {
                &::after {
                    position: absolute;
                    top: $m-border-width--l;
                    right: 0;
                    bottom: 0;
                    content: '';
                    width: $m-border-width--s;
                    background: $m-color--grey;
                }
            }
        }

        &:not(.m--is-selected):not(.m--is-disabled) {
            &[tabindex],
            a,
            button {
                transition: background $m-transition-duration ease;

                &:hover,
                &:focus {
                    background: $m-color--grey-lighter;
                }

                &:active {
                    background: $m-color--grey;
                }
            }
        }
    }
}

.m--is-skin-dark {
    .m-navbar-item {
        &.m--is-selected {
            &[tabindex],
            a,
            button {
                color: $m-color--white;
                background: $m-color--grey-dark;
            }
        }
    }
}

.m--is-skin-light {
    .m-navbar-item {
        margin: 0 $m-margin;

        &:not(.m--is-disabled) {
            &[tabindex],
            a,
            button {
                &:hover,
                &:focus {
                    color: $m-color--text;

                    &::after {
                        opacity: 1;
                    }
                }
            }
        }

        a:visited {
            color: $m-color--grey-darker;
        }

        &[tabindex],
        a,
        button {
            transition: color $m-transition-duration ease;
            color: $m-color--grey-darker;
            padding: $m-margin--s 0;

            &::after {
                transition: opacity $m-transition-duration ease;
                position: absolute;
                right: 0;
                bottom: 0;
                left: 0;
                content: '';
                width: 100%;
                height: $m-border-width--l;
                background: $m-color--grey;
                opacity: 0;
            }
        }

        &.m--is-selected {
            &[tabindex],
            a,
            button {
                color: $m-color--text;
            }
        }

        &.m--is-first {
            margin-left: 0;
        }

        &.m--is-last {
            margin-right: 0;
        }
    }
}

.m--is-skin-arrow {
    .m-navbar {
        &__list {
            padding-bottom: $m-margin--s;
        }
    }

    .m-navbar-item {
        margin: 0 $m-margin;

        &:not(.m--is-disabled) {
            &[tabindex],
            a,
            button {
                &:hover,
                &:focus {
                    color: $m-color--text;
                }
            }
        }

        a:visited {
            color: $m-color--grey-darker;
        }

        &[tabindex],
        a,
        button {
            transition: color $m-transition-duration ease;
            color: $m-color--grey-darker;
            padding: $m-margin 0;
            font-weight: $m-font-weight--bold;
        }

        &.m--is-selected {
            &[tabindex],
            a,
            button {
                color: $m-color--text;
            }
        }

        &.m--is-first {
            margin-left: 0;
        }

        &.m--is-last {
            margin-right: 0;
        }
    }
}

.m--no-margin {
    .m-navbar-item {
        margin: 0;
    }
}
