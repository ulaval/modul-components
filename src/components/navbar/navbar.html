<div class="m-navbar" :class="{ 'm--is-disabled': disabled,
              'm--is-skin-tab': skin === 'light-tab' || skin === 'dark-tab',
              'm--is-skin-light-tab': skin === 'light-tab',
              'm--is-skin-dark-tab': skin === 'dark-tab',
              'm--is-skin-light': skin === 'light',
              'm--is-skin-dark': skin === 'dark',
              'm--is-skin-plain': skin === 'plain',
              'm--is-skin-simple': skin === 'simple',
              'm--is-skin-arrow': skin === 'arrow',
              'm--no-margin': !margin }">
    <m-icon-button ref="buttonLeft" class="m-navbar__button" :skin="buttonSkin" :button-size="ComputedHeight" icon-name="left-chevron"
        @click="scrollLeft" v-if="hasScrolllH && arrowMobile"></m-icon-button>
    <nav class="m-navbar__wrap" role="menubar" ref="wrap">
        <ul class="m-navbar__list" aria-hidden="true" ref="list">
            <slot></slot>
            <li aria-hidden="true" class="m-navbar__line" :class="{ 'm--is-anim-active': animActive }" v-if="isLightSkin" ref="line"></li>
            <li aria-hidden="true" class="m-navbar__arrow" :class="{ 'm--is-anim-active': animActive }" v-if="isArrowSkin" ref="arrow"></li>
        </ul>
    </nav>
    <m-icon-button ref="buttonRight" class="m-navbar__button" :skin="buttonSkin" :button-size="ComputedHeight" icon-name="chevron"
        @click="scrollRight" v-if="hasScrolllH && arrowMobile"></m-icon-button>
</div>
