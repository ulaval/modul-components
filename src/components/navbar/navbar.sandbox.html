<div>
    <!-- Sandbox template -->
    <h2>Navbar skins</h2>
    <p class="m-u--margin-bottom">
        Show the different skins possible for the component
    </p>

    <m-button @click="toggleMultiline">{{ multiline ? 'Remove' : 'Add' }} multiline</m-button>

    <h3>Skin nav-main (default)</h3>
    <m-navbar :selected.sync="selectedItem" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant">
        <m-navbar-item value="item1">
            1 souris verte
        </m-navbar-item>
        <m-navbar-item value="item2">
            2
        </m-navbar-item>
        <m-navbar-item value="item3">
            3
        </m-navbar-item>
        <m-navbar-item value="item4">
            4
        </m-navbar-item>
        <m-navbar-item value="item5">
            5
        </m-navbar-item>
        <m-navbar-item value="item6">
            6
        </m-navbar-item>
        <m-navbar-item value="item7">
            7
        </m-navbar-item>
        <m-navbar-item value="item8">
            8
        </m-navbar-item>
        <m-navbar-item value="item9">
            9 singes
        </m-navbar-item>
        <m-navbar-item value="item10" :disabled="true">
            10 chèvres et 10 moutons
        </m-navbar-item>
        <m-navbar-item value="item11">
            11 barnacles magestueuses volant vers le lointain
        </m-navbar-item>
        <m-navbar-item value="item12">
            12 zaine d'oeufs
        </m-navbar-item>
    </m-navbar>

    <h3>Skin nav-sub</h3>
    <m-navbar skin="nav-sub" :selected.sync="selectedItem" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant">
        <m-navbar-item value="item1">
            1 souris verte
        </m-navbar-item>
        <m-navbar-item value="item2">
            2
        </m-navbar-item>
        <m-navbar-item value="item3">
            3
        </m-navbar-item>
        <m-navbar-item value="item4">
            4
        </m-navbar-item>
        <m-navbar-item value="item5">
            5
        </m-navbar-item>
        <m-navbar-item value="item6">
            6
        </m-navbar-item>
        <m-navbar-item value="item7">
            7
        </m-navbar-item>
        <m-navbar-item value="item8">
            8
        </m-navbar-item>
        <m-navbar-item value="item9">
            9 singes
        </m-navbar-item>
        <m-navbar-item value="item10" :disabled="true">
            10 chèvres et 10 moutons
        </m-navbar-item>
        <m-navbar-item value="item11">
            11 barnacles magestueuses volant vers le lointain
        </m-navbar-item>
        <m-navbar-item value="item12">
            12 zaine d'oeufs
        </m-navbar-item>
    </m-navbar>

    <h3>Skin nav-soft</h3>
    <m-navbar skin="nav-soft" :selected.sync="selectedItem" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant" style="background: #09f;">
        <m-navbar-item value="item1">
            1 souris verte
        </m-navbar-item>
        <m-navbar-item value="item2">
            2
        </m-navbar-item>
        <m-navbar-item value="item3">
            3
        </m-navbar-item>
        <m-navbar-item value="item4">
            4
        </m-navbar-item>
        <m-navbar-item value="item5">
            5
        </m-navbar-item>
        <m-navbar-item value="item6">
            6
        </m-navbar-item>
        <m-navbar-item value="item7">
            7
        </m-navbar-item>
        <m-navbar-item value="item8">
            8
        </m-navbar-item>
        <m-navbar-item value="item9">
            9 singes
        </m-navbar-item>
        <m-navbar-item value="item10" :disabled="true">
            10 chèvres et 10 moutons
        </m-navbar-item>
        <m-navbar-item value="item11">
            11 barnacles magestueuses volant vers le lointain
        </m-navbar-item>
        <m-navbar-item value="item12">
            12 zaine d'oeufs
        </m-navbar-item>
    </m-navbar>

    <h3>Skin tab-light</h3>
    <m-navbar skin="tab-light" :selected.sync="selectedItem" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant">
        <m-navbar-item value="item1">
            1 souris verte
        </m-navbar-item>
        <m-navbar-item value="item2">
            2
        </m-navbar-item>
        <m-navbar-item value="item3">
            3
        </m-navbar-item>
        <m-navbar-item value="item4">
            4
        </m-navbar-item>
        <m-navbar-item value="item5">
            5
        </m-navbar-item>
        <m-navbar-item value="item6">
            6
        </m-navbar-item>
        <m-navbar-item value="item7">
            7
        </m-navbar-item>
        <m-navbar-item value="item8">
            8
        </m-navbar-item>
        <m-navbar-item value="item9">
            9 singes
        </m-navbar-item>
        <m-navbar-item value="item10" :disabled="true">
            10 chèvres et 10 moutons
        </m-navbar-item>
        <m-navbar-item value="item11">
            11 barnacles magestueuses volant vers le lointain
        </m-navbar-item>
        <m-navbar-item value="item12">
            12 zaine d'oeufs
        </m-navbar-item>
    </m-navbar>

    <h3>Skin tab-dark</h3>
    <m-navbar skin="tab-dark" :selected.sync="selectedItem" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant">
        <m-navbar-item value="item1">
            1 souris verte
        </m-navbar-item>
        <m-navbar-item value="item2">
            2
        </m-navbar-item>
        <m-navbar-item value="item3">
            3
        </m-navbar-item>
        <m-navbar-item value="item4">
            4
        </m-navbar-item>
        <m-navbar-item value="item5">
            5
        </m-navbar-item>
        <m-navbar-item value="item6">
            6
        </m-navbar-item>
        <m-navbar-item value="item7">
            7
        </m-navbar-item>
        <m-navbar-item value="item8">
            8
        </m-navbar-item>
        <m-navbar-item value="item9">
            9 singes
        </m-navbar-item>
        <m-navbar-item value="item10" :disabled="true">
            10 chèvres et 10 moutons
        </m-navbar-item>
        <m-navbar-item value="item11">
            11 barnacles magestueuses volant vers le lointain
        </m-navbar-item>
        <m-navbar-item value="item12">
            12 zaine d'oeufs
        </m-navbar-item>
    </m-navbar>

    <h3>Skin tab-arrow</h3>
    <m-navbar skin="tab-arrow" :selected.sync="selectedItem" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant">
        <m-navbar-item value="item1">
            1 souris verte
        </m-navbar-item>
        <m-navbar-item value="item2">
            2
        </m-navbar-item>
        <m-navbar-item value="item3">
            3
        </m-navbar-item>
        <m-navbar-item value="item4">
            4
        </m-navbar-item>
        <m-navbar-item value="item5">
            5
        </m-navbar-item>
        <m-navbar-item value="item6">
            6
        </m-navbar-item>
        <m-navbar-item value="item7">
            7
        </m-navbar-item>
        <m-navbar-item value="item8">
            8
        </m-navbar-item>
        <m-navbar-item value="item9">
            9 singes
        </m-navbar-item>
        <m-navbar-item value="item10" :disabled="true">
            10 chèvres et 10 moutons
        </m-navbar-item>
        <m-navbar-item value="item11">
            11 barnacles magestueuses volant vers le lointain
        </m-navbar-item>
        <m-navbar-item value="item12">
            12 zaine d'oeufs
        </m-navbar-item>
    </m-navbar>

    <h3>Skin tab-underline</h3>
    <m-navbar skin="tab-underline" :selected.sync="selectedItem" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant">
        <m-navbar-item value="item1">
            1 souris verte
        </m-navbar-item>
        <m-navbar-item value="item2">
            2
        </m-navbar-item>
        <m-navbar-item value="item3">
            3
        </m-navbar-item>
        <m-navbar-item value="item4">
            4
        </m-navbar-item>
        <m-navbar-item value="item5">
            5
        </m-navbar-item>
        <m-navbar-item value="item6">
            6
        </m-navbar-item>
        <m-navbar-item value="item7">
            7
        </m-navbar-item>
        <m-navbar-item value="item8">
            8
        </m-navbar-item>
        <m-navbar-item value="item9">
            9 singes
        </m-navbar-item>
        <m-navbar-item value="item10" :disabled="true">
            10 chèvres et 10 moutons
        </m-navbar-item>
        <m-navbar-item value="item11">
            11 barnacles magestueuses volant vers le lointain
        </m-navbar-item>
        <m-navbar-item value="item12">
            12 zaine d'oeufs
        </m-navbar-item>
    </m-navbar>

    <h3>Skin tab-soft</h3>
    <m-navbar skin="tab-soft" :selected.sync="selectedItem" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant">
        <m-navbar-item value="item1">
            1 souris verte
        </m-navbar-item>
        <m-navbar-item value="item2">
            2
        </m-navbar-item>
        <m-navbar-item value="item3">
            3
        </m-navbar-item>
        <m-navbar-item value="item4">
            4
        </m-navbar-item>
        <m-navbar-item value="item5">
            5
        </m-navbar-item>
        <m-navbar-item value="item6">
            6
        </m-navbar-item>
        <m-navbar-item value="item7">
            7
        </m-navbar-item>
        <m-navbar-item value="item8">
            8
        </m-navbar-item>
        <m-navbar-item value="item9">
            9 singes
        </m-navbar-item>
        <m-navbar-item value="item10" :disabled="true">
            10 chèvres et 10 moutons
        </m-navbar-item>
        <m-navbar-item value="item11">
            11 barnacles magestueuses volant vers le lointain
        </m-navbar-item>
        <m-navbar-item value="item12">
            12 zaine d'oeufs
        </m-navbar-item>
    </m-navbar>

    <h2>Main navbar with sub menu</h2>
    <m-navbar skin="nav-main" :selected.sync="selectedItemMainMenu" :multiline="multiline" title-button-left="Naviguer précédent" title-button-right="Naviguer suivant">
        <m-navbar-item value="item1" @click="toggleNavPres">
            Présentation du cours
        </m-navbar-item>
        <m-navbar-item value="item2" @click="toggleNavMod">
            Modules d'apprentissage
        </m-navbar-item>
        <m-navbar-item value="item3" @click="closeNav">
            Évaluations et travaux pratiques
        </m-navbar-item>
        <m-navbar-item value="item4" @click="closeNav">
            Résultats
        </m-navbar-item>
        <m-navbar-item value="item5" @click="closeNav">
            Ressources complémentaires et références bibliographiques
        </m-navbar-item>
        <m-navbar-item value="item6" @click="closeNav">
            Calendrier de session
        </m-navbar-item>
    </m-navbar>
    <m-accordion-transition>
        <div v-if="navOpenPres" >
            <m-navbar skin="nav-sub" :selected.sync="selectedItemPres">
                <m-navbar-item value="item1">
                    Introduction
                </m-navbar-item>
                <m-navbar-item value="item2">
                    Liens entre le cours et le programme
                </m-navbar-item>
                <m-navbar-item value="item3">
                    Modalités particulières de l’offre de cours
                </m-navbar-item>
                <m-navbar-item value="item4">
                    Objectifs généraux
                </m-navbar-item>
                <m-navbar-item value="item5">
                    Objectifs particuliers
                </m-navbar-item>
                <m-navbar-item value="item6">
                    Éléments de contenu
                </m-navbar-item>
                <m-navbar-item value="item7">
                    Approche pédagogique
                </m-navbar-item>
                <m-navbar-item value="item8">
                    Modalités d'encadrement
                </m-navbar-item>
                <m-navbar-item value="item9">
                    Appréciation de l'enseignement
                </m-navbar-item>
            </m-navbar>
        </div>
    </m-accordion-transition>
    <m-accordion-transition>
        <div v-if="navOpenMod" >
            <m-navbar skin="nav-sub" :selected.sync="selectedItemMod">
                <m-navbar-item value="item1">
                    Module de la semaine 1
                </m-navbar-item>
                <m-navbar-item value="item2">
                    Module de la semaine 2
                </m-navbar-item>
                <m-navbar-item value="item3">
                    Module de la semaine 3
                </m-navbar-item>
                <m-navbar-item value="item4">
                    Module de la semaine 4
                </m-navbar-item>
                <m-navbar-item value="item5">
                    Module de la semaine 5
                </m-navbar-item>
                <m-navbar-item value="item6">
                    Module de la semaine 6
                </m-navbar-item>
                <m-navbar-item value="item7">
                    Module de la semaine 7
                </m-navbar-item>
                <m-navbar-item value="item8">
                    Module de la semaine 8
                </m-navbar-item>
                <m-navbar-item value="item9">
                    Module de la semaine 9
                </m-navbar-item>
            </m-navbar>
        </div>
    </m-accordion-transition>
</div>
