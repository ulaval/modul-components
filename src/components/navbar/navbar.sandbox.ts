import Vue, { PluginObject } from 'vue';
import { Component } from 'vue-property-decorator';

import { NAVBAR_NAME } from '../component-names';
import WithRender from './navbar.sandbox.html';

@WithRender
@Component
export class MNavbarSandbox extends Vue {
    public selectedItem = 'item4';
}

const NavbarSandboxPlugin: PluginObject<any> = {
    install(v, options): void {
        v.component(`${NAVBAR_NAME}-sandbox`, MNavbarSandbox);
    }
};

export default NavbarSandboxPlugin;
