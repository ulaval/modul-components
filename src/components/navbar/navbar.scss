@import 'abstracts/variables';

.m-navbar {
    position: relative;
    display: flex;
    justify-content: center;
    overflow: hidden;

    &.m--is-skin-dark {
        background-color: $m-color--grey-darkest;
    }

    &.m--is-skin-light {
        &::before {
            position: absolute;
            left: 0;
            right: 0;
            bottom: 0;
            height: $m-border-width--l;
            width: 100%;
            content: '';
            background: $m-color--border;
        }
    }

    &.m--is-skin-arrow {
        &::before {
            position: absolute;
            right: 0;
            bottom: $m-margin--s;
            left: 0;
            content: '';
            width: 100%;
            height: $m-border-width--s;
            background: $m-color--grey;
        }
    }

    &__button {
        position: relative;
        display: flex;
        align-items: center;
        justify-content: center;
        max-width: 64px;

        .m--is-skin-dark & {
            color: $m-color--grey-light;
            stroke-width: 2;
        }

        .m--is-anim-ready & {
            &.show-enter-active,
            &.show-leave-active {
                transition: opacity 0.3s ease, max-width 0.4s ease;
            }

            &.show-enter,
            &.show-leave-to {
                opacity: 0;
                max-width: 0;
            }
        }
    }

    &__wrap {
        position: relative;
        flex: 1 1 auto;
        overflow-x: auto;
        overflow-y: hidden;
    }

    &__list {
        position: relative;
        list-style: none;
        padding: 0;
        margin: 0;
        display: inline-flex;
        align-items: center;
        text-align: center;
        white-space: nowrap;
    }

    &__line,
    &__arrow {
        position: absolute;

        &.m--is-anim-active {
            transition: transform $m-transition-duration ease, width $m-transition-duration ease;
        }
    }

    &__line {
        bottom: 0;
        background: $m-color--active;
        height: $m-border-width--l;
    }

    &__arrow {
        bottom: $m-margin--s;

        &::before,
        &::after {
            transform: translateX(-50%);
            position: absolute;
            left: 50%;

            @include m-triangle(bottom, currentcolor, 18px);
        }

        &::before {
            color: $m-color--grey;
            top: 0;
        }

        &::after {
            color: $m-color--white;
            top: -1px;
        }
    }

    &-item {
        position: relative;
        display: flex;
        align-items: center;
        user-select: none;

        &:not(.m--is-selected):not(.m--is-disabled) {
            &,
            a,
            a:visited,
            button {
                cursor: pointer;
            }
        }

        &.m--is-disabled {
            &,
            a,
            a:visited,
            button {
                cursor: not-allowed;
            }
        }

        &,
        a,
        a:visited,
        button {
            font-family: $m-font-family;
            font-size: $m-font-size;
            font-weight: $m-font-weight;
        }

        button {
            border: none;
            background: none;
        }

        &,
        :focus {
            outline: none;
        }
    }
}
