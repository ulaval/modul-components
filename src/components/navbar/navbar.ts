import Vue, { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Prop, Watch } from 'vue-property-decorator';

import { ElementQueries } from '../../mixins/element-queries/element-queries';
import { ModulVue } from '../../utils/vue/vue';
import { NAVBAR_NAME } from '../component-names';
import NavbarItemPlugin, { BaseNavbar, Navbar } from '../navbar-item/navbar-item';
import WithRender from './navbar.html?style=./navbar.scss';

const UNDEFINED: string = 'undefined';
const PAGE_STEP: number = 4;
const THRESHOLD: number = 2.5;
const POPPER_CLASS_NAME: string = '.m-popper__popper';

export enum MNavbarSkin {
    Light = 'light',
    Dark = 'dark',
    LightTab = 'light-tab',
    DarkTab = 'dark-tab',
    Plain = 'plain',
    Simple = 'simple',
    Arrow = 'arrow'
}

@WithRender
@Component({
    mixins: [ElementQueries]
})
export class MNavbar extends BaseNavbar implements Navbar {

    @Prop()
    public selected: string;
    @Prop({
        default: MNavbarSkin.LightTab,
        validator: value =>
            value === MNavbarSkin.Light ||
            value === MNavbarSkin.Dark ||
            value === MNavbarSkin.LightTab ||
            value === MNavbarSkin.DarkTab ||
            value === MNavbarSkin.Plain ||
            value === MNavbarSkin.Simple ||
            value === MNavbarSkin.Arrow
    })
    public skin: string;
    @Prop({ default: true })
    public margin: boolean;
    @Prop()
    public disabled: boolean;
    @Prop({ default: true })
    public arrowMobile: boolean;
    @Prop({ default: false })
    public mouseover: boolean;

    private animActive: boolean = false;
    private internalValue: any | undefined = '';
    private hasScrolllH: boolean = false;
    private computedHeight: number = 0;

    public updateValue(value: any): void {
        this.model = value;
    }

    public onMouseOver(value: any, event: Event): void {
        this.$emit('mouseover', value, event);
    }

    public onClick(value: any, event: Event): void {
        this.$emit('click', value, event);
    }

    public get model(): any {
        return this.selected === undefined ? this.internalValue : this.selected;
    }

    public set model(value: any) {
        this.setAndUpdate(value);
        this.$emit('update:selected', value);
    }

    protected created(): void {
        this.internalValue = undefined;
    }

    protected mounted(): void {
        this.scrollToSelected();
        this.setupScrolllH();
        this.as<ElementQueries>().$on('resize', this.setupScrolllH);
    }

    protected beforeDestroy(): void {
        this.as<ElementQueries>().$off('resize', this.setupScrolllH);
    }

    private scrollToSelected(): void {
        this.$children.forEach(element => {
            if (element.$props.value === this.selected) {

                (this.$refs.wrap as HTMLElement).scrollLeft = element.$el.offsetLeft;

                if (this.skin === MNavbarSkin.Light) {
                    this.setPosition(element, 'line');
                }
                if (this.skin === MNavbarSkin.Arrow) {
                    this.setPosition(element, 'arrow');
                }
            }
        });
    }

    private setPosition(element, ref: string): void {
        let positionX: number = element.$el.offsetLeft;
        let width: number = element.$el.clientWidth;
        let localRef: HTMLElement = this.$refs[ref] as HTMLElement;

        localRef.style.transform = 'translate3d(' + positionX + 'px, 0, 0)';
        localRef.style.width = width + 'px';
        this.animActive = true;

    }

    @Watch('selected')
    private setAndUpdate(value): void {
        this.internalValue = value;
        if (this.skin === MNavbarSkin.Light || this.skin === MNavbarSkin.Arrow) {
            let selected: Vue | undefined = this.$children.find(element => element.$props.value === this.selected);
            if (selected) {
                this.setPosition(selected, this.skin);
            }
        }
    }

    private setupScrolllH(): void {
        let listEl: HTMLElement = this.$refs.list as HTMLElement;
        let wrapEl: HTMLElement = this.$refs.wrap as HTMLElement;
        let elComputedStyle: any = (this.$el as any).currentStyle || window.getComputedStyle(this.$el);
        let width: number = this.$el.clientWidth - parseInt(elComputedStyle.paddingLeft, 10) -
            parseInt(elComputedStyle.paddingRight, 10) - parseInt(elComputedStyle.borderLeftWidth, 10) -
            parseInt(elComputedStyle.borderRightWidth, 10);
        if (width < listEl.clientWidth) {
            this.computedHeight = listEl.clientHeight;
            this.hasScrolllH = true;
            wrapEl.style.height = this.computedHeight + 40 + 'px';
            this.$el.style.height = this.computedHeight + 'px';
        } else {
            this.hasScrolllH = false;
            wrapEl.style.removeProperty('height');
            this.$el.style.removeProperty('height');
        }
    }

    private get ComputedHeight(): string {
        return this.computedHeight + 'px';
    }

    private get buttonSkin(): string {
        return this.skin === 'dark' ? this.skin : 'light';
    }

    private get isLightSkin(): boolean {
        return this.skin === 'light';
    }

    private get isArrowSkin(): boolean {
        return this.skin === 'arrow';
    }

    private scrollLeft(event: MouseEvent): void {
        let wrapEl: HTMLElement = this.$refs.wrap as HTMLElement;
        let btnsWidth: any = ((this.$refs.buttonLeft as ModulVue).$el as HTMLElement).clientWidth + ((this.$refs.buttonRight as ModulVue).$el as HTMLElement).clientWidth;
        wrapEl.scrollLeft = wrapEl.scrollLeft - ((this.$el.clientWidth - btnsWidth) / THRESHOLD);
    }

    private scrollRight(event: MouseEvent): void {
        let wrapEl: HTMLElement = this.$refs.wrap as HTMLElement;
        let btnsWidth: any = ((this.$refs.buttonLeft as ModulVue).$el as HTMLElement).clientWidth + ((this.$refs.buttonRight as ModulVue).$el as HTMLElement).clientWidth;
        wrapEl.scrollLeft = wrapEl.scrollLeft + ((this.$el.clientWidth - btnsWidth) / THRESHOLD);
    }
}

const NavbarPlugin: PluginObject<any> = {
    install(v, options): void {
        v.prototype.$log.warn(NAVBAR_NAME + ' is not ready for production');
        v.use(NavbarItemPlugin);
        v.component(NAVBAR_NAME, MNavbar);
    }
};

export default NavbarPlugin;
