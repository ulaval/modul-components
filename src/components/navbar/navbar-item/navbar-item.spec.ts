import { createLocalVue, mount, shallow, Wrapper } from '@vue/test-utils';
import Vue, { VueConstructor } from 'vue';
import { renderComponent } from '../../../../tests/helpers/render';
import ModulPlugin from '../../../utils/modul/modul';
import { MNavbar } from '../../navbar/navbar';
import { MNavbarItem } from './navbar-item';
import NavbarItemHelper from './navbar-item-helper';

let mockIsRouterLinkActive: boolean = false;

jest.mock('./navbar-item-helper', () => ({
    isRouterLinkActive: jest.fn(() => mockIsRouterLinkActive)
}));

describe('MNavbarItem', () => {
    let wrapper: Wrapper<MNavbarItem>;
    let parentNavbar: Wrapper<MNavbar>;
    let localVue: VueConstructor<Vue>;
    beforeEach(() => {
        mockIsRouterLinkActive = false;
        Vue.use(ModulPlugin);
        localVue = createLocalVue();
        parentNavbar = shallow(MNavbar);
    });

    const initializeWrapper: (initialPropValues: any)
        => Wrapper<MNavbarItem> = (initialPropValues: any = {}) => {
            document.querySelector = () => true;
            wrapper = shallow(MNavbarItem, {
                methods: {
                    getParent(): MNavbar {
                        return parentNavbar.vm;
                    }
                },
                propsData: initialPropValues
            });

            return wrapper;
        };

    const defaultSlot: any = {
        default: `navbar item content`
    };

    it('should render correctly', () => {
        const wrapper: Wrapper<MNavbarItem> = mount(MNavbarItem, {
            localVue: localVue,
            slots: defaultSlot
        });

        expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
    });

    it('should render correctly when disabled', () => {
        const wrapper: Wrapper<MNavbarItem> = mount(MNavbarItem, {
            localVue: localVue,
            slots: defaultSlot,
            propsData: {
                disabled: true
            }
        });
        expect(wrapper.classes()).toContain('m--is-disabled');
        expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
    });

    describe('given parent Navbar has autoselect set to true', () => {
        beforeEach(() => {
            parentNavbar = shallow(MNavbar);
            parentNavbar.setProps({ autoSelect: true });
            parentNavbar.vm.updateValue = jest.fn();
        });

        it('should set parentNavbar value if current item is router-link-active', () => {
            mockIsRouterLinkActive = true;
            jest.spyOn(parentNavbar.vm, 'updateValue');
            jest.spyOn(NavbarItemHelper, 'isRouterLinkActive');

            initializeWrapper({ value: 'someValue' });

            expect(NavbarItemHelper.isRouterLinkActive).toHaveBeenCalledWith(wrapper.vm);
            expect(parentNavbar.vm.updateValue).toHaveBeenCalledWith('someValue');
        });

        it('should not set parentNavbar value if current item is not router-link-active', () => {
            mockIsRouterLinkActive = false;
            jest.spyOn(parentNavbar.vm, 'updateValue');
            jest.spyOn(NavbarItemHelper, 'isRouterLinkActive');

            initializeWrapper({ value: 'someValue' });

            expect(NavbarItemHelper.isRouterLinkActive).toHaveBeenCalledWith(wrapper.vm);
            expect(parentNavbar.vm.updateValue).not.toHaveBeenCalled();
        });
    });

    describe('given parent Navbar has autoselect set to false', () => {
        beforeEach(() => {
            parentNavbar = shallow(MNavbar);
            parentNavbar.setProps({ autoSelect: false });
        });

        it('should not set parentNavbar value if current item is router-link-active', () => {
            mockIsRouterLinkActive = true;
            jest.spyOn(parentNavbar.vm, 'updateValue');
            jest.spyOn(NavbarItemHelper, 'isRouterLinkActive');

            initializeWrapper({ value: 'someValue' });

            expect(NavbarItemHelper.isRouterLinkActive).not.toHaveBeenCalledWith(wrapper.vm);
            expect(parentNavbar.vm.updateValue).not.toHaveBeenCalled();
        });

        it('should not set parentNavbar value if current item is not router-link-active', () => {
            mockIsRouterLinkActive = false;
            jest.spyOn(parentNavbar.vm, 'updateValue');
            jest.spyOn(NavbarItemHelper, 'isRouterLinkActive');

            initializeWrapper({ value: 'someValue' });

            expect(NavbarItemHelper.isRouterLinkActive).not.toHaveBeenCalledWith(wrapper.vm);
            expect(parentNavbar.vm.updateValue).not.toHaveBeenCalled();
        });
    });
});
