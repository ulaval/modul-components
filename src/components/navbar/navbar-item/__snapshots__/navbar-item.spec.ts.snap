// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MNavbarItem should render correctly 1`] = `
<li role="menuitem" class="m-navbar-item">
    <div tabindex="0" role="button" class="m-navbar-item__contents"><span class="m-navbar-item__text">navbar item content</span></div>
</li>
`;

exports[`MNavbarItem should render correctly when disabled 1`] = `
<li role="menuitem" class="m-navbar-item m--is-disabled">
    <div tabindex="-1" class="m-navbar-item__contents"><span class="m-navbar-item__text">navbar item content</span></div>
</li>
`;
